src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -149,10 +149,10 @@ private void AdvanceTo(BufferSegment consumedSegment, int consumedIndex, BufferS
             // Remove all blocks that are freed (except the last one)
             while (returnStart != returnEnd)
             {
-                BufferSegment next = returnStart.NextSegment;
+                BufferSegment? next = returnStart.NextSegment;
                 returnStart.ResetMemory();
                 ReturnSegmentUnsynchronized(returnStart);
-                returnStart = next;
+                returnStart = next!;
```

Comment:
I think this would be better if you made `next` non-nullable and then put the `!` on `returnStart.NextSegment`.  The loop is based on the idea that NextSegment will not be null as long as returnStart != returnEnd, so it makes sense to express that explicitly with the `!` in the right place.  It also has the benefit of reducing the number of annotations needed.

Reply:
Sure, thanks!